### PR TITLE
Add mock libs for testing, improve code coverage for sunpy.util.util

### PR DESCRIPTION
**Pull request by m-charlton**
- Added support for 'mocking' tests

- Code coverage for `sunpy.util.util` increased from 96% -> 100%

- Minor changes to ensure files pass `pep8`

- Altered docstrings to conform to dev standard

I've started to use mocks for the test framework. This has involved adding [mock](https://github.com/testing-cabal/mock),
and [pytest-mock](https://pypi.python.org/pypi/pytest-mock).

All works as expected in my local environment. So I grep'd the code base for all
occurrences of `pytest`, to give an idea about which files need to be changed.

Came up with the following:

- `.rtd.environment` added `pytest-mock`, should `hypothesis` be added too?

- `.travis.yml` added `mock` & `pytest-mock`

- `tox.ini` added `mock` & `pytest-mock`. The `hypothesis` package seems to be missing, also
  there are sections for python 2.6, should these be removed?

- `appveyor.yml` added `mock` & `pytest-mock`

- `doc/source/guide/installation/advanced.rst`: Added links for `hypothesis`, `mock`, 
  `pytest-cov` & `pytest-mock`, though *not* a description of what these packages do.

- `doc/source/requirements.txt` updated, though this file seems a duplicate of
  `requirements/base.txt` & `requirements/extra.txt`  

- `requirements/extra.txt` added `hypothesis`, `mock` & `pytest-mock`

Changed documentation built and proof-read.

Tested against py2.7 & py3.5.

May have missed some files/entries out or added entries where they should not be.
If this is the case, then please correct me.

### CODE REVIEW

**Comment by m-charlton** [comment]
Fails as CONDA DEPENDENCIES cannot find pytest-mock

```
ImportError: No module named 'pytest_mock'
```

Investigating

**Comment by m-charlton** [comment]
Should `pytest-mock` be in `PIP_DEPENDENCIES` instead of `CONDA_DEPENDENCIES`?

**Comment by Cadair** [comment]
if there is no conda package for it (which it looks like there isn't) then yes. Also you shouldn't need to modify `rtd-environment.yml` as RTD will not run the tests.

**Comment by m-charlton** [comment]
Cheers, I'll 'backout' changes to `.rtd.environment.yml`, change `.travis.yml` and resubmit.

**Comment by m-charlton** [comment]
I think, I *might* have found the problem. In the travis logs: 

```
...
$ if [[ $SETUP_CMD == *figure* ]]; then conda env create --file sunpy/tests/figure_tests_env_$PYTHON_VERSION.yml; source activate sunpy-figure-tests-$PYTHON_VERSION; fi
Fetching package metadata ...............
Solving package specifications: .
expat-2.1.0-0. 100% || Time: 0:00:00  22.62 MB/s
...pip-9.0.1-py35 100% || Time: 0:00:00  51.67 MB/s
pytest-3.0.5-p 100% || Time: 0:00:00  34.84 MB/s
...
dbus post-link :: /etc/machine-id not found ..
dbus post-link :: .. using /proc/sys/kernel/random/boot_id 
...
```

Looking at `sunpy/tests/figure_tests_env_2.7.yml` & `sunpy/tests/figure_tests_env_3.5.yml`
the package dependencies are explicitly listed.

My fault, I must have missed then when I did the first grep.

It looks as if conda forge does build [pytest-mock](https://anaconda.org/conda-forge/pytest-mock)
so I'll move pytest-mock back to the CONDA_DEPENDENCIES and carry out the following 
changes:

`sunpy/tests/figure_tests_env_2.7.yml`

Add `mock` and `pytest-mock`


`sunpy/tests/figure_tests_env_3.5.yml`

Add `pytest-mock`

File: sunpy/tests/figure_tests_env_3.5.yml [hunk 1]
```diff
@@ -46,6 +46,7 @@ dependencies:
 - pyparsing=2.1.4=py35_0
 - pyqt=5.6.0=py35_2
 - pytest=3.0.5=py35_0
+- pytest-mock=1.5.0=py35_0
 - python=3.5.2=0
```
Comment (dpshelio) [inline sunpy/tests/figure_tests_env_3.5.yml]: Is it `mock` missing here?

Comment (m-charlton) [inline sunpy/tests/figure_tests_env_3.5.yml]: From Py3.3 mock comes as part of the `unittest.mock` package. Only have to add it
in for the 2.7.yml file

**Comment by m-charlton** [comment]
Thanks!

**Pull request merged**